Automagically propagate modular JDK options when JDK9+ is detected

ropalka opened the pull request:
This PR addresses the following JIRAs:

https://issues.jboss.org/browse/WFCORE-3962 (partially)
https://issues.jboss.org/browse/WFCORE-4000
https://issues.jboss.org/browse/WFCORE-4001
https://issues.jboss.org/browse/WFCORE-4039
https://issues.jboss.org/browse/WFLY-10811

ropalka commented:
@bstansberry @rsvoboda - send me your feedback guys.
This PR supersedes PRs:
https://github.com/wildfly/wildfly-core/pull/3420
https://github.com/wildfly/wildfly/pull/11499

rsvoboda commented:
@marekkopecky please take a look too

I will look into it tomorrow.

rsvoboda commented:
@jamezp could be also interested as he is commenting on related PR https://github.com/wildfly/wildfly-core/pull/340

bstansberry commented on process-controller/src/main/java/org/jboss/as/process/Main.java:
```diff
@@ -221,7 +222,9 @@ public JBossThreadFactory run() {
-        initialCommand.add(jvmName);
+        final JdkType jdkType = JdkType.createFromJavaExecutable(jvmName);
+        initialCommand.add(jdkType.getJavaExecutable());
+        initialCommand.addAll(jdkType.getDefaultArguments());
```
Is this necessary? The PC and the HC use the same VM.  And generally they share JAVA_OPTS as well, with domain.conf just providing a hook for setting up a jdwp only on one or the other.

We shouldn't be overriding user provided options with defaults (e.g. we think --illegal-access=permit is great but they do not and explicitly say so.) But checking for that here is starting to get complicated and ProcessController needs to be as simple as possible.

ropalka replied:
Of course it is not necessary @bstansberry - removing it.

ropalka commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/management/ServerStartFailureTestCase.java:
```diff
@@ -117,16 +117,16 @@ public static void tearDownDomain() throws Exception {
-        waitUntilState(client, "slave", "failure-two", "FAILED");
-        waitUntilState(client, "slave", "failure-three", "FAILED");
+        waitUntilState(client, "slave", "failure-two", "DISABLED");
+        waitUntilState(client, "slave", "failure-three", "DISABLED");
```
Hi @jamezp , different output is caused by fact JDK definition in domain configs points to not existing location: /usr/java/wrongPath
This PR validates JDK path - thus it is causing FAILED -> DISABLED change.

ropalka commented on version/src/main/java/org/wildfly/jdk/version/JdkType.java:
```diff
@@ -0,0 +1,170 @@
+/*
```
WFCORE "server" maven subproject cannot see WFCORE "host-controller" subproject.
I didn't want to introduce "server" -> "host-controller" dependency, so
I analyzed "host-controller", "process-controller" and "server" maven subproject dependencies
and only "version" is available to all of them. What other location would you suggest @jamezp and @bstansberry ?

bstansberry replied:
I don't see any use of this in the 'server' module. My thought was you did this because you used it in 'process-controller' but I believe you are removing that.

If it's needed in 'server' it could go in server as 'host-controller' already depends on 'server'.

ropalka replied:
I moved it to "process-controller" module @bstansberry

marekkopecky commented on host-controller/src/main/java/org/jboss/as/host/controller/model/jvm/JvmOptionsBuilderFactory.java:
```diff
@@ -24,186 +24,124 @@
+    }
 
-        // This builder doesn't override anything currently, since the addPermgen behavior
-        // became standard. But we keep the concept in case we do vm-specific stuff
```
@bstansberry @jmesnil Are you ok with removing of this concept?

ropalka replied:
The AddPermGen mechanism was doing nothing so I removed it.

bstansberry replied:
Fine with me. It's an internal implementation detail and the implementation long ago drifted away from the original need for IbmJvmOptionsBuilder.

ropalka commented on host-controller/src/main/java/org/jboss/as/host/controller/model/jvm/JvmOptionsBuilderFactory.java:
```diff
@@ -24,186 +24,124 @@
+        this.jdkType = jdkType;
     }
 
-    public static JvmOptionsBuilderFactory getInstance() {
```
Correct @marekkopecky

bstansberry replied:
Definitely correct.

rsvoboda commented:
> Embedded CLI use-case is not fixed by this PR. 

Unless `--add-modules=java.se` problem is solved in JBoss Modules I thing the only way how to solve embedded CLI use-case is by adding `--add-modules=java.se` to java process which runs such code.

Embedded CLI use-case is also valid for installer where users on Java 11 will need to run `java --add-modules=java.se -jar eap-installer.jar` but users on Java 8 will need to run just `java -jar eap-installer.jar`

rsvoboda commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/management/ServerStartFailureTestCase.java:
```diff
@@ -136,10 +136,16 @@ public void testDomainLifecycleMethods() throws Throwable {
-        result = validateResponse(startServer(client, "slave", "failure-three"));
-        Assert.assertThat(result.asString(), is("FAILED"));
+        ModelNode result = validateFailedResponse(startServer(client, "slave", "failure-two"));
+        String failureDescription = result.asString();
```
result.asString() value changed here too, similar as above - comments by James and Brian?

marekkopecky replied:
@ropalka Any update here?

bstansberry replied:
@rsvoboda I agree; this work should not change the semantics of this operation.  Like the other test change I commented on above, it's arguable that the change is valid, but this PR isn't the place to change it. IIRC the op returning the status of the server goes way, way back.

jfdenise commented on core-feature-pack/src/main/resources/content/bin/domain.bat:
```diff
@@ -99,6 +99,29 @@ if "x%JAVA_HOME%" == "x" (
   )
 )
 
+"%JAVA%" --add-modules=java.se -version >nul 2>&1 && set MODULAR_JDK=true || set MODULAR_JDK=false
```
In order to have it work in jconsole.bat, I had to add parenthesis when setting MODULAR_JDK var.
&& (set MODULAR_JDK=true) || (set MODULAR_JDK=false)

kabir commented:
@ropalka it also needs changes in the ps1 files - In case it helps @jfdenise did some work in https://github.com/wildfly/wildfly/pull/11550

marekkopecky commented on core-feature-pack/src/main/resources/content/bin/domain.sh:
```diff
@@ -96,6 +96,26 @@ if [ "x$JAVA" = "x" ]; then
     fi
 fi
 
+$JAVA --add-modules=java.se -version > /dev/null 2>&1 && MODULAR_JDK=true || MODULAR_JDK=false
```
MODULAR_JDK variable is not used in domain.sh, so domain is unable to start with JDK8:
```
[mkopecky@dhcp-10-40-4-197 bin]$ ./domain.sh 
=========================================================================

  JBoss Bootstrap Environment

  JBOSS_HOME: /home/mkopecky/erase8/wildfly

  JAVA: /home/mkopecky/base/jdk1.8.0_162/bin/java

  JAVA_OPTS: -server -Xms64m -Xmx512m -XX:MaxMetaspaceSize=256m -Djava.net.preferIPv4Stack=true -Djboss.modules.system.pkgs=org.jboss.byteman -Djava.awt.headless=true --add-exports=java.base/sun.nio.ch=ALL-UNNAMED --add-exports=jdk.unsupported/sun.misc=ALL-UNNAMED --add-exports=jdk.unsupported/sun.reflect=ALL-UNNAMED --add-modules=java.se

=========================================================================

Unrecognized option: --add-exports=java.base/sun.nio.ch=ALL-UNNAMED
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
[mkopecky@dhcp-10-40-4-197 bin]$ 
```

marekkopecky commented:
appclient is not fixed on jdk11:
```
[mkopecky@dhcp-10-40-4-197 bin]$ ./appclient.sh 
org.jboss.modules.ModuleNotFoundException: java.se
	at org.jboss.modules.Module.addPaths(Module.java:1266)
	at org.jboss.modules.Module.link(Module.java:1622)
	at org.jboss.modules.Module.relinkIfNecessary(Module.java:1650)
	at org.jboss.modules.ModuleLoader.loadModule(ModuleLoader.java:296)
	at org.jboss.modules.Main.main(Main.java:437)
[mkopecky@dhcp-10-40-4-197 bin]$ 
```

marekkopecky commented:
jdr is not fixed as well: 
```
[mkopecky@dhcp-10-40-4-197 bin]$ ./jdr.sh 
org.jboss.modules.ModuleNotFoundException: java.se
	at org.jboss.modules.Module.addPaths(Module.java:1266)
	at org.jboss.modules.Module.link(Module.java:1622)
	at org.jboss.modules.Module.relinkIfNecessary(Module.java:1650)
	at org.jboss.modules.ModuleLoader.loadModule(ModuleLoader.java:296)
	at org.jboss.modules.Main.main(Main.java:437)
[mkopecky@dhcp-10-40-4-197 bin]$ 
```

marekkopecky commented:
domain.bat is not correct with JDK8:
```
C:\Users\Administrator\playground\wfly.25.ropalka.1\wildfly\bin>domain.bat
Calling "C:\Users\Administrator\playground\wfly.25.ropalka.1\wildfly\bin\domain.conf.bat"
Setting JAVA property to "C:\Users\Administrator\playground\jdk1.8.0_181\bin\java"
===============================================================================

  JBoss Bootstrap Environment

  JBOSS_HOME: "C:\Users\Administrator\playground\wfly.25.ropalka.1\wildfly"

  JAVA: "C:\Users\Administrator\playground\jdk1.8.0_181\bin\java"

  JAVA_OPTS: "-Xms64M -Xmx512M -XX:MaxMetaspaceSize=256m -Djava.net.preferIPv4Stack=true -Djboss.modules.system.pkgs=org.jboss.byteman -server"

===============================================================================

07:04:10,392 INFO  [org.jboss.modules] (main) JBoss Modules version 1.8.5.Final
07:04:10,751 INFO  [org.jboss.threads] (main) JBoss Threads version 2.3.2.Final
Exception in thread "main" java.lang.IllegalStateException: WFLYPC0024: Could not find java executable under C:\Users\Administrator\playground\jdk1.8.0_181\bin\java.
        at org.jboss.as.process.jdk.JdkType.createFromJavaExecutable(JdkType.java:118)
        at org.jboss.as.process.Main.start(Main.java:225)
        at org.jboss.as.process.Main.main(Main.java:77)
        at org.jboss.modules.Module.run(Module.java:352)
        at org.jboss.modules.Module.run(Module.java:320)
        at org.jboss.modules.Main.main(Main.java:593)
```

jamezp commented:
Maybe we should have a `bin/modular-jdk.sh` which will do the testing and set a `$MODULAR_JDK` environment variable so we can cut down on the duplication in the scripts. This way scripts in full can also take advantage of the same logic. Essentially something similar to what we do with the `*.conf` files.

n1hility commented on process-controller/src/main/java/org/jboss/as/process/jdk/JdkType.java:
```diff
@@ -0,0 +1,151 @@
+        if (!javaExecutable.exists()) {
+            throw ProcessLogger.ROOT_LOGGER.cannotFindJavaExe(javaBinDir.getAbsolutePath());
+        }
+        return new JdkType(new File(javaHomeDir, JMODS_DIR).exists(), javaHomeDir.getAbsolutePath(), javaExecutable.getAbsolutePath());
```
Could you look into (if you haven't already) whether this FS based detection scheme is reliable for other JVMs we support (IBM etc), and if there is any variance on where this directory might exist (e.g. installation options someone might pick).

n1hility commented on process-controller/src/main/java/org/jboss/as/process/jdk/JdkType.java:
```diff
@@ -0,0 +1,151 @@
+        if (!javaExecutable.exists()) {
+            throw ProcessLogger.ROOT_LOGGER.cannotFindJavaExe(javaBinDir.getAbsolutePath());
+        }
+        return new JdkType(new File(javaHomeDir, JMODS_DIR).exists(), javaHomeDir.getAbsolutePath(), javaExecutable.getAbsolutePath());
```
Could you look into (if you haven't already) whether this FS based detection scheme is reliable for other JVMs we support (IBM etc), and if there is any variance on where this directory might exist (e.g. installation options someone might pick).

bstansberry replied:
I'm concerned that what we're doing here is imposing constraints on how people configure things.

For example, what if --default-jvm=/usr/bin/java or something like that? Or /bin/java or /lib/java (both of which exist on my laptop).  In the existing code, we're not imposing constraints on the value of --default-jvm, but in this PR we do because JdkType.createFromJavaExecutable wants to walk from its value up two levels in order to find "java home" and then look for jmods under it.

The --default-jvm value is user controllable in domain.sh via the $JAVA env var.

n1hility replied:
Ok this is what I was worried about, that the layout is unreliable. I think the solution for standalone.* should be mirrored in domain mode. Before launching the real JVM, a leading detection execution using the --add-module + -version error code trick could be performed.

jamezp commented:
Added the fixme logo as the `domain.sh` is broken with JDK 8.

bstansberry commented on process-controller/src/main/java/org/jboss/as/process/jdk/JdkType.java:
```diff
@@ -0,0 +1,151 @@
+import org.jboss.as.process.logging.ProcessLogger;
+
+/**
+ * Jdk type detection utility.
```
s/Jdk/JVM

bstansberry commented on process-controller/src/main/java/org/jboss/as/process/jdk/JdkType.java:
```diff
@@ -0,0 +1,151 @@
+ *
+ * @author <a href="mailto:ropalka@redhat.com">Richard Opalka</a>
+ */
+public final class JdkType {
```
This should be called JvmType, as we do not require a JDK to run.

bstansberry commented:
The last commit at https://github.com/wildfly/wildfly-core/compare/master...bstansberry:server-start-failure fixes the issue with ServerStartFailureTestCase.

bstansberry commented on process-controller/src/main/java/org/jboss/as/process/Main.java:
```diff
@@ -221,7 +222,8 @@ public JBossThreadFactory run() {
         final List<String> initialCommand = new ArrayList<String>();
-        initialCommand.add(jvmName);
+        final JdkType jdkType = JdkType.createFromJavaExecutable(jvmName);
+        initialCommand.add(jdkType.getJavaExecutable());
```
The changes to this file are not necessary.  There hasn't been a problem with the existing code and the new code here isn't doing anything relevant for JDK 11 support.

Eliminating this also mean JdkType can go in the same package as JvmOptionsBuilderFactory.

bstansberry replied:
Removing this fixes the test failures on Windows. https://github.com/wildfly/wildfly-core/pull/3478

bstansberry commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/management/ServerStartFailureTestCase.java:
```diff
@@ -136,10 +136,16 @@ public void testDomainLifecycleMethods() throws Throwable {
+        failureDescription = result.asString();
+        Assert.assertTrue(failureDescription.contains("IllegalStateException"));
+        Assert.assertTrue(failureDescription.contains("Java home"));
+        Assert.assertTrue(failureDescription.contains("does not exist"));
```
@jamezp This will be removed and the test will be unchanged.

ropalka commented:
Hi guys,

   everything should be fixed now.

@marekkopecky I revisited WFCORE shell scripts and they should be functional now. Could you please verify Power Shell & Batch scripts on Windows platform? If there'll be some issue could you send a fix PR against this branch: https://github.com/ropalka/wildfly-core/tree/modular-jdk

@jamezp I refactored modular jdk detection code to common.(sh,bat,ps1) scripts as you suggested.

@bstansberry I used your commit in this PR. I refactored JdkType to JvmType and moved it from "process-controller" to "host-controller" maven subproject.

@n1hility modular jvm detection is now based on spawning new java process and detecting its error code.

CC-ing @rsvoboda just FYI.

ropalka commented:
Correct @marekkopecky . I'm working on WFLY scripts now.

jamezp commented on core-feature-pack/src/main/resources/content/bin/domain.sh:
```diff
@@ -96,6 +98,12 @@ if [ "x$JAVA" = "x" ]; then
+setDefaultModularJvmOptions $PROCESS_CONTROLLER_JAVA_OPTS
+PROCESS_CONTROLLER_JAVA_OPTS="$PROCESS_CONTROLLER_JAVA_OPTS $DEFAULT_MODULAR_JVM_OPTIONS"
+setDefaultModularJvmOptions $HOST_CONTROLLER_JAVA_OPTS
+HOST_CONTROLLER_JAVA_OPTS="$HOST_CONTROLLER_JAVA_OPTS $DEFAULT_MODULAR_JVM_OPTIONS"
```
This needs to be tested, but it may need to be invoke later. Essentially once all the JAVA_OPTS are filled.

jamezp commented on host-controller/src/main/java/org/jboss/as/host/controller/ManagedServerBootCmdFactory.java:
```diff
@@ -174,6 +177,9 @@ public ManagedServerBootCmdFactory(final String serverName, final ModelNode doma
         this.sslContextSupplier = createSSLContextSupplier(serverModel, expressionResolver);
+        // Defer initializing the vm type until we know we are at a point where a VM config
+        // validation failure 1) is acceptable and 2) can be handled properly
+        //this.jdkType = getJdkType();
```
Should this be removed?

jamezp commented on host-controller/src/main/java/org/jboss/as/host/controller/jvm/JvmType.java:
```diff
@@ -0,0 +1,193 @@
+        assert javaExecutable != null;
+
+        if (javaExecutable.equals(JAVA_WIN_EXECUTABLE) || javaExecutable.equals(JAVA_UNIX_EXECUTABLE)) {
+            return createFromEnvironmentVariable(forLaunch);
```
I'm not sure if this is correct. What if the `java` executable lives in the `$PATH`, but the `JAVA_HOME` isn't set or the system property is set to a different JVM? Maybe that's not possible, but if it is just something to consider.

jamezp commented on core-feature-pack/src/main/resources/content/bin/add-user.sh:
```diff
@@ -1,12 +1,15 @@
 #!/bin/sh
 
+. ./common.sh
```
This needs to happen after the `$DIRNAME` is configured and should be `. "$DIRNAME/common.sh`. Without that it can't be launched from any other directory than bin, e.g. `cd $JBOSS_HOME/bin && ./add-user.sh`.

jamezp commented on core-feature-pack/src/main/resources/content/bin/domain.sh:
```diff
@@ -1,5 +1,7 @@
 #!/bin/sh
 
+. ./common.sh
+
```
This needs to happen after the `$DIRNAME` is configured and should be `. "$DIRNAME/common.sh`. Without that it can't be launched from any other directory than bin, e.g. `cd $JBOSS_HOME/bin && ./domain.sh`.

jamezp commented on core-feature-pack/src/main/resources/content/bin/jboss-cli.sh:
```diff
@@ -1,6 +1,9 @@
 #!/bin/sh
 
+. ./common.sh
+
```
This needs to happen after the `$DIRNAME` is configured and should be `. "$DIRNAME/common.sh`. Without that it can't be launched from any other directory than bin, e.g. `cd $JBOSS_HOME/bin && ./jboss-cli.sh`.

jamezp commented on core-feature-pack/src/main/resources/content/bin/standalone.sh:
```diff
@@ -1,5 +1,7 @@
 #!/bin/sh
 
+. ./common.sh
+
```
This needs to happen after the `$DIRNAME` is configured and should be `. "$DIRNAME/common.sh`. Without that it can't be launched from any other directory than bin, e.g. `cd $JBOSS_HOME/bin && ./standalone.sh`.

jamezp commented on core-feature-pack/src/main/resources/content/bin/vault.sh:
```diff
@@ -1,5 +1,7 @@
 #!/bin/sh
 
+. ./common.sh
+
```
This needs to happen after the `$DIRNAME` is configured and should be `. "$DIRNAME/common.sh`. Without that it can't be launched from any other directory than bin, e.g. `cd $JBOSS_HOME/bin && ./vault.sh`.

jamezp commented on host-controller/src/main/java/org/jboss/as/host/controller/jvm/SecurityUtils.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * @author <a href="mailto:ropalka@redhat.com">Richard Opalka</a>
+ */
+final class SecurityUtils {
```
Is there a reason not to use the `WildFlySecurityManager`?

jamezp commented:
@ropalka @marekkopecky My guess is for the full scripts we need a core release and merge into full so that the `common.(sh|bat|ps1)` scripts are included.

ropalka commented:
I fixed bash and batch scripts to be location unaware @jamezp - good catch!
SecurityUtils was replaced with WildFlySecurityManager as you proposed.
@marekkopecky could you confirm please that windows batch and power shell scripts are functional?
They should work from $JBOSS_HOME/bin directory as well as outside of that directory.

marekkopecky commented:
@ropalka All scripts don't work on linux:

```
[mkopecky@dhcp-10-40-4-197 bin]$ jdk11
[mkopecky@dhcp-10-40-4-197 bin]$ ./add-user.sh 
./add-user.sh: line 54: setDefaultModularJvmOptions: command not found
org.jboss.modules.ModuleNotFoundException: java.se
	at org.jboss.modules.Module.addPaths(Module.java:1266)
	at org.jboss.modules.Module.link(Module.java:1622)
	at org.jboss.modules.Module.relinkIfNecessary(Module.java:1650)
	at org.jboss.modules.ModuleLoader.loadModule(ModuleLoader.java:296)
	at org.jboss.modules.Main.main(Main.java:437)
[mkopecky@dhcp-10-40-4-197 bin]$ jdk8
[mkopecky@dhcp-10-40-4-197 bin]$ ./add-user.sh 
./add-user.sh: line 54: setDefaultModularJvmOptions: command not found

What type of user do you wish to add? 
 a) Management User (mgmt-users.properties) 
 b) Application User (application-users.properties)
(a): 
```

marekkopecky commented:
@ropalka All powershell scripts don't work as well. You can check JDK8 cli&standalone logs, but I can attach more logs:
```
PS C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\bin> .\jboss-cli.ps1
Get-Default-Modular-Jvm-Options : Cannot bind argument to parameter 'opts' because it is null.
At C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\bin\common.ps1:147 char:69
+ ... T_MODULAR_JVM_OPTS = Get-Default-Modular-Jvm-Options -opts $JAVA_OPTS
+                                                                ~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Get-Default-Modular-Jvm-Options], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Get-Default-Modular-Jvm-Options

You are disconnected at the moment. Type 'connect' to connect to the server or 'help' for the list of supported commands.
[disconnected /]
PS C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\bin> .\standalone.ps1
=================================================================================

  JBoss Bootstrap Environment

  JBOSS_HOME: C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly

  JAVA: C:\Users\Administrator\playground\jdk1.8.0_181\bin\java.exe

  MODULE_OPTS:

  JAVA_OPTS: -Xms64M -Xmx512M -XX:MetaspaceSize=96M -XX:MaxMetaspaceSize=256m -Djava.net.preferIPv4Stack=true -Djboss.modules.system.pkgs=org.jboss.byteman  -Dorg.jboss.boot.log.file=C:\Users\
Administrator\playground\wfly.25.ropalka.2\wildfly\standalone\log/server.log -Dlogging.configuration=file:C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\standalone\configuration/l
ogging.properties -Djboss.home.dir=C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly -Djboss.server.base.dir=C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\standalone -D
jboss.server.config.dir=C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\standalone\configuration

=================================================================================

Start-WildFly-Process : Cannot bind argument to parameter 'programArguments' because it is an empty string.
At C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\bin\standalone.ps1:40 char:41
+ Start-WildFly-Process -programArguments $PROG_ARGS -runInBackground $ ...
+                                         ~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Start-WildFly-Process], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorEmptyStringNotAllowed,Start-WildFly-Process

PS C:\Users\Administrator\playground\wfly.25.ropalka.2\wildfly\bin>
```

ropalka commented:
I'm installing Microsoft server on my laptop to fix batch & power shell issues.

jamezp commented:
Superseded by #3488. The new PR should address the PowerShell scripts.
